modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C

```diff
@@ -0,0 +1,170 @@
+  params.addParam<Real>(
+      "signal_variance", 1, "Signal Variance (sigma_f^2) to use for kernel calculation.");
+  params.addParam<Real>(
+      "noise_variance", 0, "Noise Variance (sigma_n^2) to use for kernel calculation.");
+  params.addParam<Real>("gamma", "Gamma to use for Exponential Covariance Kernel");
+  params.addParam<unsigned int>("p", "Integer p to use for Matern Hald Integer Covariance Kernel");
+  params.addParam<bool>(
+      "standardize_params", true, "Standardize (center and scale) training parameters (x values)");
+  params.addParam<bool>(
+      "standardize_data", true, "Standardize (center and scale) training data (y values)");
+
+  return params;
+}
+
+GaussianProcessTrainer::GaussianProcessTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _kernel_type(getParam<MooseEnum>("kernel_function")),
+    _training_params(declareModelData<RealEigenMatrix>("_training_params")),
+    _training_params_mean(declareModelData<RealEigenVector>("_training_params_mean")),
+    _training_params_var(declareModelData<RealEigenVector>("_training_params_var")),
+    _training_data(declareModelData<RealEigenMatrix>("_training_data")),
+    _training_data_mean(declareModelData<RealEigenVector>("_training_data_mean")),
+    _training_data_var(declareModelData<RealEigenVector>("_training_data_var")),
+    _K(declareModelData<RealEigenMatrix>("_K")),
+    _K_results_solve(declareModelData<RealEigenMatrix>("_K_results_solve")),
+    _covar_function(
+        declareModelData<std::unique_ptr<CovarianceFunction::CovarianceKernel>>("_covar_function")),
+    _standardize_params(getParam<bool>("standardize_params")),
+    _standardize_data(getParam<bool>("standardize_data"))
+{
+}
+
+void
+GaussianProcessTrainer::initialize()
+{
+}
+
+void
+GaussianProcessTrainer::initialSetup()
+{
+
+  // Results VPP
+  _values_distributed = isVectorPostprocessorDistributed("results_vpp");
+  _values_ptr = &getVectorPostprocessorValue(
+      "results_vpp", getParam<std::string>("results_vector"), !_values_distributed);
+
+  // Sampler
+  _sampler = &getSamplerByName(getParam<SamplerName>("sampler"));
+  _n_params = _sampler->getNumberOfCols();
+
+  // Check if sampler dimension matches number of distributions
+  std::vector<DistributionName> dname = getParam<std::vector<DistributionName>>("distributions");
+  if (dname.size() != _n_params)
+    mooseError("Sampler number of columns does not match number of inputted distributions.");
+
+  _covar_function = CovarianceFunction::makeCovarianceKernel(_kernel_type, this);
+}
+
+void
+GaussianProcessTrainer::execute()
+{
+  // Check if results of samples matches number of samples
+  __attribute__((unused)) dof_id_type num_rows =
+      _values_distributed ? _sampler->getNumberOfLocalRows() : _sampler->getNumberOfRows();
+  mooseAssert(num_rows == _values_ptr->size(),
+              "Sampler number of rows does not match number of results from vector postprocessor.");
```

Comment:
You should put this in `initialize` and use `paramError` instead. See PolynomailRegressionTrainer for an example. I haven't gotten around to fixing this for PolynomialChaos